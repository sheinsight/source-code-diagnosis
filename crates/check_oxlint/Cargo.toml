[package]
name = "check_oxlint"
version = "0.1.0"
edition = "2024"
 

[dependencies]
anyhow = { workspace = true }
napi-derive = { workspace = true }
napi = { workspace = true }
utils = { path = "../utils" }
beans = { path = "../beans" }
insta = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
pathdiff = { workspace = true }

oxc_linter = { workspace = true }
oxc = { workspace = true }
