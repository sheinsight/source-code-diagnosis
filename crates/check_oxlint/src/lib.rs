use std::{fmt::Display, rc::Rc, sync::Arc};

use napi_derive::napi;
use oxc_diagnostics::Severity;
use oxc_linter::{AllowWarnDeny, FixKind, LintFilter, LinterBuilder, Oxlintrc};
use oxc_semantic::SemanticBuilder;
use serde::Serialize;
use serde_json::json;

#[napi(object)]
#[derive(Debug, Clone, Serialize)]
pub struct CheckOxlintLabelsSpan {
  pub offset: u32,
  pub length: u32,
}

#[napi(object)]
#[derive(Debug, Clone, Serialize)]
pub struct CheckOxlintLabelsResponse {
  pub span: CheckOxlintLabelsSpan,
}

#[napi(object)]
#[derive(Debug, Clone, Serialize)]
pub struct CheckOxlintResponse {
  pub file_name: String,
  pub help: String,
  pub url: String,
  pub severity: String,
  pub code: String,
  pub message: String,
  pub labels: Vec<CheckOxlintLabelsResponse>,
}

impl Display for CheckOxlintResponse {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", serde_json::to_string_pretty(self).unwrap())
  }
}

pub fn check_oxlint(
  rules: Vec<&'static str>,
  args: utils::GlobArgs,
) -> anyhow::Result<Vec<CheckOxlintResponse>> {
  let responses = utils::glob_by(
    |path| {
      let builder = utils::SemanticBuilder::with_file(path).unwrap();

      let parse = oxc_parser::Parser::new(
        &builder.allocator,
        &builder.source_code,
        builder.source_type,
      )
      .parse();

      let program = builder.allocator.alloc(parse.program);

      let semantic = SemanticBuilder::new()
        .with_check_syntax_error(false)
        .with_cfg(true)
        .build(program);

      let module_record = Arc::new(oxc_linter::ModuleRecord::new(
        &path,
        &parse.module_record,
        &semantic.semantic,
      ));

      // let rule: Vec<&'static str> = vec![
      //   "no-debugger",
      //   "no-console",
      //   "constructor-super",
      //   "for-direction",
      //   "getter-return",
      //   "no-async-promise-executor",
      //   "no-case-declarations",
      //   "no-class-assign",
      //   "no-compare-neg-zero",
      //   "no-cond-assign",
      //   "no-const-assign",
      //   "no-constant-binary-expression",
      //   "no-constant-condition",
      //   "no-control-regex",
      //   "no-delete-var",
      //   "no-dupe-class-members",
      //   "no-dupe-else-if",
      //   "no-dupe-keys",
      //   "no-duplicate-case",
      //   "no-empty",
      //   "no-empty-character-class",
      //   "no-empty-pattern",
      //   "no-ex-assign",
      //   "no-fallthrough",
      //   "no-func-assign",
      //   "no-global-assign",
      //   "no-import-assign",
      //   "no-inner-declarations",
      //   "no-invalid-regexp",
      //   "no-irregular-whitespace",
      //   "no-loss-of-precision",
      //   "no-new-native-nonconstructor",
      //   "no-nonoctal-decimal-escape",
      //   "no-obj-calls",
      //   "no-prototype-builtins",
      //   "no-redeclare",
      //   "no-regex-spaces",
      //   "no-self-assign",
      //   "no-setter-return",
      //   "no-shadow-restricted-names",
      //   "no-sparse-arrays",
      //   "no-this-before-super",
      //   "no-unexpected-multiline",
      //   "no-unreachable",
      //   "no-unsafe-finally",
      //   "no-unsafe-negation",
      //   "no-unsafe-optional-chaining",
      //   "no-unused-labels",
      //   "no-useless-catch",
      //   "no-useless-escape",
      //   "use-isnan",
      //   "valid-typeof",
      // ];

      // let filter: Vec<LintFilter> = rules
      //   .iter()
      //   .map(|r| LintFilter::new(AllowWarnDeny::Deny, r.to_string()).unwrap())
      //   .collect::<Vec<_>>();

      // let rules_str = rules
      //   .iter()
      //   .map(|item| format!(r#""{}": "error""#, item))
      //   .collect::<Vec<_>>()
      //   .join(",");

      let rules: serde_json::Value = rules
        .iter()
        .map(|&rule| {
          (
            rule.to_string(),
            serde_json::Value::String("error".to_string()),
          )
        })
        .collect::<serde_json::Map<String, serde_json::Value>>()
        .into();

      let config: Oxlintrc = serde_json::from_value(json!({
          // "plugins": ["react", "oxc"],
          "rules": rules,
          "globals":{
            "ROOT_REDUCER": "readonly",
            "ROOT_ROUTER": "readonly",
          }
      }))
      .unwrap();
      let linter = LinterBuilder::from_oxlintrc(true, config)
        .with_fix(FixKind::None)
        .build();

      // let linter = LinterBuilder::empty()
      //   .with_filters(filter)
      //   .with_fix(FixKind::None)
      //   .build();

      let semantic = Rc::new(semantic.semantic);

      let diagnostics = linter.run(path, semantic, module_record);

      let responses = diagnostics
        .into_iter()
        .map(|d| {
          let error = d.error;

          let file_name = pathdiff::diff_paths(path, &args.cwd)
            .unwrap()
            .display()
            .to_string();

          let help = error
            .help
            .as_ref()
            .map(|h| h.to_string())
            .unwrap_or("".to_string());
          let url = error
            .url
            .as_ref()
            .map(|u| u.to_string())
            .unwrap_or("".to_string());

          let severity = match error.severity {
            Severity::Advice => "advice".to_string(),
            Severity::Warning => "warning".to_string(),
            Severity::Error => "error".to_string(),
          };

          let labels = error
            .labels
            .as_ref()
            .map(|v| {
              v.iter()
                .map(|l| CheckOxlintLabelsResponse {
                  span: CheckOxlintLabelsSpan {
                    offset: l.offset() as u32,
                    length: l.len() as u32,
                  },
                })
                .collect::<Vec<_>>()
            })
            .unwrap_or_default();

          return CheckOxlintResponse {
            file_name,
            help,
            url,
            severity,
            code: error.code.to_string(),
            message: error.message.to_string(),
            labels: labels,
          };
        })
        .collect::<Vec<_>>();

      Some(responses)
    },
    &args,
  )?;

  Ok(responses.into_iter().flatten().collect())
}
