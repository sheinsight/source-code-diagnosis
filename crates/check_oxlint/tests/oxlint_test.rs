use check_oxlint::check_oxlint;
use std::env::current_dir;

macro_rules! test_oxlint {
  ($name:ident, $rule:expr) => {
    #[test]
    fn $name() -> anyhow::Result<()> {
      let curr = current_dir()?.join("tests").join("features").join($rule);

      let js_args = utils::GlobJsArgs {
        pattern: None,
        ignore: None,
        cwd: curr.to_string_lossy().to_string(),
      };

      let args = utils::GlobArgs::from(js_args);
      let mut responses = check_oxlint(
        format!(
          r#"{{
        "rules":{{
          "{}":"error"
        }}
      }}"#,
          $rule
        ),
        args,
      )?;

      responses.sort_by(|a, b| a.file_name.cmp(&b.file_name));

      responses.iter().for_each(|response| {
        insta::assert_snapshot!(response.to_string());
      });

      Ok(())
    }
  };
}

test_oxlint!(test_debugger, "no-debugger");
test_oxlint!(test_console, "no-console");

test_oxlint!(test_constructor_super, "constructor-super");

test_oxlint!(test_for_direction, "for-direction");

test_oxlint!(test_getter_return, "getter-return");

test_oxlint!(test_no_async_promise_executor, "no-async-promise-executor");

test_oxlint!(
  test_no_constant_binary_expression,
  "no-constant-binary-expression"
);

test_oxlint!(test_no_case_declarations, "no-case-declarations");

test_oxlint!(test_no_class_assign, "no-class-assign");

test_oxlint!(test_no_compare_neg_zero, "no-compare-neg-zero");

test_oxlint!(test_no_cond_assign, "no-cond-assign");

test_oxlint!(test_no_const_assign, "no-const-assign");

test_oxlint!(test_no_constant_condition, "no-constant-condition");

test_oxlint!(test_no_control_regex, "no-control-regex");

test_oxlint!(test_no_delete_var, "no-delete-var");

test_oxlint!(test_no_dupe_class_members, "no-dupe-class-members");

test_oxlint!(test_no_dupe_else_if, "no-dupe-else-if");

test_oxlint!(test_no_dupe_keys, "no-dupe-keys");

test_oxlint!(test_no_duplicate_case, "no-duplicate-case");

test_oxlint!(test_no_new_for_builtins, "unicorn/new-for-builtins");
