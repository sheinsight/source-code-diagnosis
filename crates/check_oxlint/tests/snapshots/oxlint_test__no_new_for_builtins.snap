---
source: crates/check_oxlint/tests/oxlint_test.rs
expression: response.to_string()
snapshot_kind: text
---
{
  "file_name": "new-for-builtins.js",
  "help": "",
  "url": "https://oxc.rs/docs/guide/usage/linter/rules/unicorn/new-for-builtins.html",
  "severity": "error",
  "code": "eslint-plugin-unicorn(new-for-builtins)",
  "message": "Use `new Array()` instead of `Array()`",
  "labels": [
    {
      "span": {
        "offset": 13,
        "length": 9
      }
    }
  ]
}
